crypto/bn/bn_gcd.c

```diff
@@ -621,3 +515,108 @@ static BIGNUM *BN_mod_inverse_no_branch(BIGNUM *in,
     bn_check_top(ret);
     return ret;
 }
+
+/*-
+ * This function is based on the constant-time GCD work by Bernstein and Yang:
+ * https://eprint.iacr.org/2019/266
+ * Generalized fast GCD function to allow even inputs.
+ * The algorithm first finds the shared powers of 2 between
+ * the inputs, and removes them, reducing at least one of the
+ * inputs to an odd value. Then it proceeds to calculate the GCD.
+ * Before returning the resulting GCD, we take care of adding
+ * back the powers of two removed at the beginning.
+ * Note: we assume the bit length of both inputs is public information,
+ * since access to top potentially leaks this information.
```

Comment:
We discussed this internally and it seems ok to write down this caveat here and probably also in the documentation for this function.

@mattcaswell @davidben do you think it is acceptable to potentially leak bn->top or should we devise something to access the whole allocated array and just discard computation on words after top (assuming inputs for which word-length is confidential already preallocate bn->d to a public length)?

Reply:
Good question.  Fixing this everywhere once and for all might be better than doing it here.